### ISSUE TITLE
[LFRic] Fortran Kernel generation should capture module information

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#393]
At the moment, LFRic Fortran Kernel generation only captures the kernel subroutine information, not the surround module and any declarations contained within.

**Comment by rupertford** [comment]
There is an issue with doing this. The Kernel Node is currently used to generate a Kernel call for the PSy-layer back-end and we would want to use it to also generate a module for the Kernel back-end.

some options are

a) give the back-end some boolean flags to decide what to do

b) try to use the context i.e. is it the first node encountered or not

c) use a different Fortran back-end for PSy-layer and Kernel generation (we could subclass the Kernel back-end to make the PSy-layer back-end and overwrite the Kernel method?)

d) split a Kernel into two types of Node - a KernelCall and a Kernel.

c) and d) sound the most promising to me but as I'm not sure where we're going to go with the PSy-layer Nodes and generation in the future, I might be missing something. Any thoughts @sergisiso or @arporter.

**Comment by TeranIvy** [comment]
And I would also like the Kernel Node to know about its arguments and their attributes (e.g. data type, kind, array shape) for generating Psyke driver :) Kernel Call doesn't help as it doesn't have full information that need to be written/read. That information is scattered in various classes (mainly DynInvoke and DynCollection) and it would be really useful to have it one place.

It seems to me that common elements for Kernel Generation, Kernel Call, Kernel Driver and Kernel Stub are a) kernel arguments and their attributes and b) module information. Kernel Call/Driver/Stub also use kernel module information (metadata) to derive argument list and their attributes required for declarations (PSy layer, kernel driver and kernel stub, respectively) and kernel calls (PSy layer and kernel driver, respectively). I presume names and attributes for Kernel Generation are obtained from directly from the parsed kernel code.
The common method to all four classes is argument declaration. Kernel Call will add PSy layer call, Kernel Driver will add write/read statements, Kernel Generation will add other statements from subroutine body, etc. Another common method is use of module information although the specific implementation will differ for each subclass (e.g. generation of USE statements, kernel arguments).

To summarise, I think it would be good to have a parent class with common data (arguments and their attributes) and common methods (argument declaration, use of kernel module information). Subclasses will implemented specifics and add other things they need.

**Comment by TeranIvy** [comment]
There is issue #364 for restructuring Kernel class hierarchy, although that issue only deals with the difference between in-lined kernels and kernel calls. I guess that in-lined kernels will still require at least two of the common methods listed above (defining argument names and attributes; code generation). How about laying some design foundations before proceeding with any of these issues?

**Comment by TeranIvy** [comment]
| Properties  | Call | Extract |  Stub | Generation | Transformation |
| ------------ | ---- | ------- | ----- | ------------ | -------------- |
| **Layer modified** | PSy, Alg | PSy, Driver | Kernel Stub (independent) | Kernel (& PSy?) |  Kernel (& PSy?) |
| **Argument attributes from** | Algorithm (invoke call), Kernel Module (metadata), API rules | Algorithm (invoke call), Kernel Module (metadata), API rules | Kernel Module (metadata), API rules | Kernel Module (declarations, global variables) | Kernel Module (?) |
| **Generate argument declarations**  | Yes (PSy) | Yes (PSy, Driver) | Yes (Stub) | Yes (Kernel) | ? |
| **Generate Kernel calls**  | Yes (PSy) | Yes (PSy, Driver) | No | No (?) | No (?) |
| **KernelSchedule required/modified**  | No | No | No | Yes | Yes (?) |
| **Program unit** | Module | Module, Program | Module | Module (?) | Module (?) |
| **Frontend(s)**  | fparser (1 & 2) | fparser (1 & 2) | fparser 1 (?) | fparser 2 (?) | fparser 2 (?) |
| **Intermediate representation**  | PSyIR (?) | PSyIR (?) | PSyIR (?) | PSyIR (?) | PSyIR (?) |
| **Backend(s)**  | Fortran | Fortran | Fortran | Fortran (and ?) | Fortran (and ?) |

**Comment by TeranIvy** [comment]
> 
> ... I'm not sure where we're going to go with the PSy-layer Nodes and generation in the future ...

It would be good to talk about how the same basic structures are being used for code analysis and generation in other layers beside PSy - since they were originally developed with only PSy layer in mind.  I think we need a bit of generalisation here:
- `Node` - a representative unit of an element in a `Schedule`.
- `Schedule` - a sequence of `Node`s enclosed within a well-defined region with a set of rules on how they can be ordered/arranged.

"Representative" means having (at least access to) the required information (data and methods) to generate the desired element of code. Elements vary with their purpose, e.g. a branch of an `if` construct in a `KernelSchedule`, Kernel Call in an `InvokeSchedule`. Some elements will be applicable throughout layers: e.g. `Loop`s, OpenMP transformations (can be applied to `Loop`s in `Kernel` and `Invoke` Schedules).

Regarding the definition of a `Schedule`, please note that this does not automatically imply belonging only to a top-level structure like a `Kernel` or an `Invoke` as we have now. For instance, a well-defined region can be:
- `if` construct formed from `if` statement branches within a `KernelSchedule` with a general rule that an individual `if` branch cannot be moved outside of the construct,
- A PSy-layer region within an `InvokeSchedule` (e.g. profile, extract, ACC data region) with local rules on allowed `Node`s and transformations that can be applied to them. 

This definition of a `Schedule` would make easier to modify/generate regions of the top-level structures, provided that individual `Node`s within have the required information for code generation.

It is also worth thinking about how different layers relate to each other - i.e. through what constructs. An obvious one is a `KernelNode`, since it can generate code in different layers depending what it is used for.

**Comment by rupertford** [comment]
To recap. Transformations on LFRic kernels modify the Kernel PSyIR. Currently there is only one transformation available - DynKernelConstTrans. However, the PSyIR only represents the original kernel subroutine and does not keep any information about the kernel module containing the subroutine. This means that when using the Fortran back-end to translate from PSyIR to Fortran there is no way to know 1) if there was a surrounding module and what its name was, 2) if there were any other statements associated with the module (use statements, variable declarations, metadata, other subroutines). At the moment, the PSyiR assumes a kernel module name derived from the kernel subroutine name and always declares r_def and i_def via a use statement (from the constants_mod module). 

The long term (language independent) solution for this seems to be to have the PSyIR also capture module information. This would require the FParser2ASTProcessor to be modified so that it's generate_schedule() method (which takes an fp2 parse tree of a module and a subroutine name as input) captures any module information in addition to the subroutine.

I was wondering whether a short term fix would be to write out the original fp2 module code but replace the original subroutine with the modified PSyIR subroutine. This would have the advantage of keeping all of the information not captured by the PSyIR (e.g. other kernels, data statements etc.). However, I think that it would mean that modified PSyIR kernels could only be written out to Fortran (or at least modified PSyIR kernels which contain important declaration information in module scope), which I think is too much of a hack, given that we are developing an OpenCL backend.

**Comment by rupertford** [comment]
We will need to capture the original module name in the kernel somehow so that the Fortran back-end can re-create it if required.

One option would be to use an extension of the annotations support. Using the annotations support as-is would be hacky as the annotations are expected to be strings, not name/value pairs. I suppose that an additional properties list or perhaps language specific dictionary could be added.

Another option would be to capture the concept of subroutines/methods being contained within something as a language independent container and to name the container with the module name. Fortran modules are essentially static classes so this approach would fit well if we ever wanted to capture classes and methods. I think it would also work for files with multiple subroutines/procedures in them and would allow us to capture data that is global over subroutines/procedures but constrained by the scope of the container (e.g. file scope variables, class variables and module variables).

I think I like the latter solution as it supports the concept of a collection of method/subroutines/procedures/functions which is found in many languages.

**Comment by sergisiso** [comment]
A few comments:

- Wait until #510 is on master to make any changes :)
- A you mention, `generate_schedule` already gets the module fparser2 tree, so getting the information is doable.
-  "Another option would be to capture the concept of subroutines/methods". I think that is something we should do as not all subroutines are kernels. And not all kernels are subroutines (nemo). But at the moment CodedKern<->KernelSchedule seems the caller<->callee implementation.
- 
> I was wondering whether a short term fix would be to write out the original fp2 module code but replace the original subroutine with the modified PSyIR subroutine. This would have the advantage of keeping all of the information not captured by the PSyIR (e.g. other kernels, data statements etc.). However, I think that it would mean that modified PSyIR kernels could only be written out to Fortran (or at least modified PSyIR kernels which contain important declaration information in module scope), which I think is too much of a hack, given that we are developing an OpenCL backend.

That is the same as a CodeBlock but outwards (we have the same limitations in OpenCL with CodeBlock and I think it is alright, we can still reason about the code and maybe find a way around). Maybe we should add this outwards CodeBlock concept (e.g. CodeHorizon ?)

**Comment by rupertford** [comment]
Here is an email I sent to @sergisiso and @arporter ...

> I've been thinking about capturing module information in the PSyIR and I've come to the conclusion that a container holding a collection of kernels (or perhaps holding a mixture of containers and kernels) which can have a symbol table, is a relatively general (language independent) concept. It is essentially a way of supporting a hierarchical namespace (which also allows the scoping of variables.
> 
> Of course this would allow Fortran modules to map nicely into the PSyIR and situations where we have more than one kernel in a module, which is why I am thinking about it, but I think it could potentially be useful for C, where variables can be declared with file scope, and perhaps for classes, where methods are contained within a class.
> 
> I guess one issue is that it is not necessary if there is only one kernel in a container and another issue is that in this case there would be 2 ways to represent a kernel (with and without a container).
> 
> In terms of implementation I would create a container node which would reference 1 or more kernels (and/or containers?), would have a symbol table and would have a name. It would also have an annotation with "from_module" in our case. A kernel would then have an optional reference to a container.
> 
> What do you think?
> 

We discussed this proposal this morning and have decided to go ahead. We agreed that the container annotation is not needed for LFRic and that a kernel should always have a container i.e. it should not be optional.

As suggested by @sergisiso, I will wait until #510 is on master until creating a branch.

**Comment by rupertford** [comment]
Created branch 433_container_node. This will create the PSyIR Node that will be used to capture module information.

**Comment by rupertford** [comment]
Having just discussed Containers with @sergisiso and @arporter it has been decided that containers and kernelschedules should be children of a container and that the names of these containers and kernelschedules should be kept in the SymbolTable i.e. the SymbolTable should be extended to support "function" names as well as variable names.

**Comment by rupertford** [comment]
The remainder of this issue (that is not addressed in pr #518) has been captured in issue #520.
